hyperspy/drawing/_widgets/line2d.py

```diff
@@ -0,0 +1,386 @@
+# (at your option) any later version.
+#
+#  HyperSpy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with  HyperSpy.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import numpy as np
+
+from hyperspy.drawing.widgets import ResizableDraggableWidgetBase
+
+
+def unit_vector(vector):
+    """ Returns the unit vector of the vector.  """
+    return vector / np.linalg.norm(vector)
+
+
+def angle_between(v1, v2):
+    """ Returns the angle in radians between @D vectors 'v1' and 'v2'::
+
+            >>> angle_between((1, 0), (0, 1))
+            1.5707963267948966
+            >>> angle_between((1, 0), (1, 0))
+            0.0
+            >>> angle_between((1, 0), (-1, 0))
+            3.141592653589793
+    """
+    v1_u = unit_vector(v1)
+    v2_u = unit_vector(v2)
+    angle = np.arctan2(v2_u[1], v2_u[0]) - np.arctan2(v1_u[1], v1_u[0])
+    # angle = np.arccos(np.dot(v1_u, v2_u))
+    if np.isnan(angle):
+        if (v1_u == v2_u).all():
+            return 0.0
+        else:
+            return np.pi
+    return angle
+
+
+class Line2DWidget(ResizableDraggableWidgetBase):
+
+    """A free-form line on a 2D plot. Enables dragging and moving the end
+    points, but also allows rotation of the widget by moving the mouse beyond
+    the end points of the line.
+
+    The widget adds the 'linewidth' attribute, which is different from the size
+    in the following regards: 'linewidth' is simply the width of the patch
+    drawn from point to point. If 'size' is greater than 1, it will in
+    principle select a rotated rectangle. If 'size' is greater than 4, the
+    bounds of this rectangle will be visualized by two dashed lines along the
+    outline of this rectangle, instead of a signle line in the center.
+
+    The widget also adds the attributes 'radius_resize', 'radius_move' and
+    'radius_rotate' (defaults: 5, 5, 10), which determines the picker radius
+    for resizing, aka. moving the edge points (by picking within
+    'radius_resize' from an edge point); for moving (by picking within
+    'radius_move' from the body of the line); and for rotation (by picking
+    within 'radius_rotate' of the edge points on the "outside" of the line).
+    The priority is in the order resize, rotate, move; so the 'radius_rotate'
+    should always be larger than 'radius_resize' if the function is to be
+    accessible (putting it lower is an easy way to disable the functionality).
+
+
+    NOTE: This widget's internal position does not lock to axes points by
+          default.
+    NOTE: The 'position' is now a 2D tuple: tuple(tuple(x1, x2), tuple(y1, y2))
+    NOTE: The 'size' property corresponds to line width, so it has a len() of
+    only one.
```

Comment:
Due to this point, `_do_snap_size` should be rewritten for this class - at the moment it assumes `size` has as many dimensions as there are axes